File: pymath/norms_metrics.py
from .math_utils import newton_sqrt


def vector_l1_norm(v):
    """
    Compute the L1 norm (Manhattan norm) of a vector.

    Args:
        v (iterable): Input vector.

    Returns:
        float: L1 norm of the vector.
    """
    return sum(abs(x) for x in v)


def vector_l2_norm(v):
    """
    Compute the L2 norm (Euclidean norm) of a vector.

    Args:
        v (iterable): Input vector.

    Returns:
        float: L2 norm of the vector.
    """
    return newton_sqrt(sum(x ** 2 for x in v))


def vector_linf_norm(v):
    """
    Compute the L-infinity norm (maximum absolute value) of a vector.

    Args:
        v (iterable): Input vector.

    Returns:
        float: L-infinity norm of the vector.
    """
    return max(abs(x) for x in v)


def matrix_frobenius_norm(matrix):
    """
    Compute the Frobenius norm of a matrix.

    Args:
        matrix (iterable of iterables): Input matrix.

    Returns:
        float: Frobenius norm of the matrix.
    """
    return newton_sqrt(sum(sum(x ** 2 for x in row) for row in matrix))


def matrix_l1_norm(matrix):
    """
    Compute the L1 norm (maximum column sum) of a matrix.

    Args:
        matrix (iterable of iterables): Input matrix.

    Returns:
        float: L1 norm of the matrix.
    """
    return max(sum(abs(row[i]) for row in matrix) for i in range(len(matrix[0])))


def matrix_linf_norm(matrix):
    """
    Compute the L-infinity norm (maximum row sum) of a matrix.

    Args:
        matrix (iterable of iterables): Input matrix.

    Returns:
        float: L-infinity norm of the matrix.
    """
    return max(sum(abs(x) for x in row) for row in matrix)


def euclidean_distance(v1, v2):
    """
    Compute the Euclidean distance between two vectors.

    Args:
        v1 (iterable): First vector.
        v2 (iterable): Second vector.

    Returns:
        float: Euclidean distance between v1 and v2.
    """
    return newton_sqrt(sum((x - y) ** 2 for x, y in zip(v1, v2)))


def manhattan_distance(v1, v2):
    """
    Compute the Manhattan distance between two vectors.

    Args:
        v1 (iterable): First vector.
        v2 (iterable): Second vector.

    Returns:
        float: Manhattan distance between v1 and v2.
    """
    return sum(abs(x - y) for x, y in zip(v1, v2))


def chebyshev_distance(v1, v2):
    """
    Compute the Chebyshev distance between two vectors.

    Args:
        v1 (iterable): First vector.
        v2 (iterable): Second vector.

    Returns:
        float: Chebyshev distance between v1 and v2.
    """
    return max(abs(x - y) for x, y in zip(v1, v2))

File: src/norms_metrics.cpp
#include "norms_metrics.hpp"
#include "math_utils.hpp"

namespace norms_metrics {

    double vector_l1_norm(const std::vector<double>& v) {
        double sum = 0;
        for (double x : v) {
            sum += std::abs(x);
        }
        return sum;
    }

    double vector_l2_norm(const std::vector<double>& v) {
        double sum = 0;
        for (double x : v) {
            sum += x * x;
        }
        return math_utils::newton_sqrt(sum);
    }

    double vector_linf_norm(const std::vector<double>& v) {
        return *std::max_element(v.begin(), v.end(), [](double a, double b) {
            return std::abs(a) < std::abs(b);
        });
    }

    double matrix_frobenius_norm(const std::vector<std::vector<double>>& matrix) {
        double sum = 0;
        for (const auto& row : matrix) {
            for (double x : row) {
                sum += x * x;
            }
        }
        return math_utils::newton_sqrt(sum);
    }

    double matrix_l1_norm(const std::vector<std::vector<double>>& matrix) {
        double max_sum = 0;
        size_t n_cols = matrix[0].size();
        for (size_t i = 0; i < n_cols; ++i) {
            double col_sum = 0;
            for (const auto& row : matrix) {
                col_sum += std::abs(row[i]);
            }
            max_sum = std::max(max_sum, col_sum);
        }
        return max_sum;
    }

    double matrix_linf_norm(const std::vector<std::vector<double>>& matrix) {
        double max_sum = 0;
        for (const auto& row : matrix) {
            double row_sum = 0;
            for (double x : row) {
                row_sum += std::abs(x);
            }
            max_sum = std::max(max_sum, row_sum);
        }
        return max_sum;
    }

    double euclidean_distance(const std::vector<double>& v1, const std::vector<double>& v2) {
        double sum = 0;
        for (size_t i = 0; i < v1.size(); ++i) {
            sum += (v1[i] - v2[i]) * (v1[i] - v2[i]);
        }
        return math_utils::newton_sqrt(sum, 1e-10);
    }

    double manhattan_distance(const std::vector<double>& v1, const std::vector<double>& v2) {
        double sum = 0;
        for (size_t i = 0; i < v1.size(); ++i) {
            sum += std::abs(v1[i] - v2[i]);
        }
        return sum;
    }

    double chebyshev_distance(const std::vector<double>& v1, const std::vector<double>& v2) {
        double max_diff = 0;
        for (size_t i = 0; i < v1.size(); ++i) {
            max_diff = std::max(max_diff, std::abs(v1[i] - v2[i]));
        }
        return max_diff;
    }

}
